src/artemis/devices/oav/oav_centring.py

```diff
@@ -0,0 +1,692 @@
+                tip_y_positions,
+            ) = tuple(
+                self.get_spacial_data_from_pin_at_different_rotations(rotation_points)
+            )
+
+            (
+                index_of_largest_width,
+                indices_orthogonal_to_largest_width,
+            ) = self.find_widest_point_and_orthogonal_point(
+                x_positions, y_positions, widths, omega_angles
+            )
+
+            x = x_positions[index_of_largest_width]
+            y = y_positions[index_of_largest_width]
+            best_omega_angle = omega_angles[index_of_largest_width]
+
+            # get the angle sufficiently orthogonal to the best omega and
+            # store its mid line - which will be the magnitude in the z axis on 90 degree rotation
+            index_orthogonal_to_largest_width = indices_orthogonal_to_largest_width[-1]
+            best_omega_angle_orthogonal = omega_angles[
+                index_orthogonal_to_largest_width
+            ]
+            z = mid_lines[index_orthogonal_to_largest_width][x]
+
+            # we need to store the tips of the angles orthogonal-ish to the best angle
+            orthogonal_tips_x = tip_x_positions[indices_orthogonal_to_largest_width]
+
+            # best_omega_angle_90 could be zero, which used to cause a failure - d'oh!
+            if best_omega_angle_orthogonal is None:
+                LOGGER.error("Unable to find loop at 2 orthogonal angles")
+                return
+
+            # extract the max_tip_distance again as it could have changed
+            self.oav_parameters.max_tip_distance = (
+                self.oav_parameters._extract_dict_parameter(
+                    "max_tip_distance", reload_json=True
+                )
+            )
+            # extract the microns per pixel of the zoom level of the camera
+            self.oav_parameters.load_microns_per_pixel(str(self.oav_parameters.zoom))
+
+            # get the max tip distance in pixels
+            max_tip_distance_pixels = (
+                self.oav_parameters.max_tip_distance
+                / self.oav_parameters.micronsPerXPixel
+            )
+
+            # get the average distance between the tips
+            tip_x = np.median(orthogonal_tips_x)
+
+            # if x exceeds the max tip distance then set it to the max tip distance
+            tip_distance_pixels = x - tip_x
+            if tip_distance_pixels > max_tip_distance_pixels:
+                x = max_tip_distance_pixels + tip_x
+
+            # get the scales of the image in microns
+            x_size, y_size = (pv.obj.value for pv in self.get_sizes_from_pvs())
+            x_scale, y_scale = self.get_scale(x_size, y_size)
+            x_move, y_move = self.calculate_beam_distance_in_microns(
+                int(x * x_scale), int(y * y_scale)
+            )
+            x_y_z_move = self.microns_to_xyz_move(x_move, y_move, best_omega_angle)
+
+            # if we've succeeded and it's the last run then set the x_move_and y_move
```

Comment:
Haven't changed this since GDA conversion. It's weird... see when GDA gets the orthogonal rotation data, they intentionally don't filter out 0 waveforms - they say `# deliberately not gon_omega_array2`. As a consequence of this, if the widest bulge data is functional, but the samples orthogonal to it are duds, z can be 0.

 I've messaged Mark about this, personally I think when we find orthogonal angles we should use the filtered arrays. The orthogonal angles being duds (all 0) is exceptional behaviour which should result in failure.